Frameworks/CoreGraphics/CGContextCairo.mm

```diff
@@ -1887,17 +1896,49 @@
-    CGAffineTransform transform = CGAffineTransformConcat(textTransform, userTransform);
-    imgRenderTarget->SetTransform(D2D1::Matrix3x2F(transform.a, transform.b, transform.c, transform.d, transform.tx, transform.ty));
+    // DWrite will crash if we try to give it glyphs that are below this threshold
+    // Though this value is approximate, it is small enough to not be noticeable while still safe
+    const float threshold = 0.5f;
+    if ((fabs(finalTransform.a * glyphRun->fontEmSize) <= threshold && fabs(finalTransform.b * glyphRun->fontEmSize) <= threshold) ||
+        (fabs(finalTransform.c * glyphRun->fontEmSize) <= threshold && fabs(finalTransform.d * glyphRun->fontEmSize) <= threshold)) {
+        TraceWarning(TAG, L"Glyphs too small to be rendered");
+        return;
+    }
+
+    // Set the brush color to the current values from the context.
+    D2D1::ColorF brushColor =
+        D2D1::ColorF(curState->curFillColor.r, curState->curFillColor.g, curState->curFillColor.b, curState->curFillColor.a);
 
+    imgRenderTarget->BeginDraw();
     // Draw the glyph using ID2D1RenderTarget
     ComPtr<ID2D1SolidColorBrush> brush;
     THROW_IF_FAILED(imgRenderTarget->CreateSolidColorBrush(brushColor, &brush));
-    imgRenderTarget->DrawGlyphRun(D2D1::Point2F(0, 0), glyphRun, brush.Get(), DWRITE_MEASURING_MODE_NATURAL);
+    if (textTransform.a == 1.0f && fabs(textTransform.d) == 1.0f && textTransform.b == 0.0f && textTransform.c == 0.0f) {
+        // Apply the two transforms giving us the final result
+        imgRenderTarget->SetTransform(
+            D2D1::Matrix3x2F(finalTransform.a, finalTransform.b, finalTransform.c, finalTransform.d, finalTransform.tx, finalTransform.ty));
+        imgRenderTarget->DrawGlyphRun(D2D1::Point2F(0, 0), glyphRun, brush.Get(), DWRITE_MEASURING_MODE_NATURAL);
+    } else {
+        // Text scaling and rotation apply to each glyph relative to its origin, so we must draw each glyph transformed independently
+        DWRITE_GLYPH_RUN individualGlyphRun{ glyphRun->fontFace,
+                                             glyphRun->fontEmSize,
+                                             1, // Since this is glyph by glyph, glyphCount is one
+                                             glyphRun->glyphIndices,
+                                             nullptr,
+                                             nullptr,
+                                             glyphRun->isSideways,
+                                             glyphRun->bidiLevel };
+        for (uint32_t i = 0; i < glyphRun->glyphCount; ++i, ++(individualGlyphRun.glyphIndices)) {
+            CGAffineTransform transform = CGAffineTransformConcat(textTransform, userTransform);
+            imgRenderTarget->SetTransform(D2D1::Matrix3x2F(transform.a, transform.b, transform.c, transform.d, transform.tx, transform.ty));
+            imgRenderTarget->DrawGlyphRun(D2D1::Point2F(0, 0), &individualGlyphRun, brush.Get(), DWRITE_MEASURING_MODE_NATURAL);
```

Comment:
i like this, but call it out in a comment

---
In reply to: [91789032](https://github.com/Microsoft/WinObjC/pull/1520#discussion_r91789032) [](ancestors = 91789032)